### PR TITLE
[Train] Add pydantic as a dependency for `ray[train]` and `ray[tune]`

### PR DESCRIPTION
**Pull request by woshiyyya**
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/lightning/_lightning_utils.py
Comment (matthewdeng) [inline python/ray/train/lightning/_lightning_utils.py:1]: Rebase?

File: python/setup.py [hunk 1]
```diff
@@ -266,7 +266,14 @@ def get_packages(self):
...
             "watchfiles",
         ],
-        "tune": ["pandas", "tensorboardX>=1.9", "requests", pyarrow_dep, "fsspec"],
+        "tune": [
+            "pandas",
+            "tensorboardX>=1.9",
+            "requests",
+            pyarrow_dep,
+            "fsspec",
+            "pydantic!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.*,!=2.4.*,<3",
```
Comment (matthewdeng) [inline python/setup.py:275]: I don't think we need `pydantic` as a dependency of `train`, just `default`.

**Review by matthewdeng — CHANGES_REQUESTED** [review]

File: python/setup.py (hunk 1 above)
Comment (woshiyyya) [reply python/setup.py:275]: I am planning to put the pydantic class under `ray/train/_internal`. If we don't have it for `train`, are we still be able to build `ray[train]`?

Comment (matthewdeng) [reply python/setup.py:275]: The code should just be guarded by whether or not `pydantic` is installed. Similar to `torch`. 😄

**Pull request closed without merging**